### PR TITLE
Initial substitution algorithm

### PR DESCRIPTION
**Pull request by mariari**

### CODE REVIEW

File: src/Juvix/Core/IR/Subst.hs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+data T primTy primVal
+  = T
+      { _depth :: Natural,
+        _sub :: Map.T Key (Types.Elim primTy primVal)
+      }
+  deriving (Show)
+
+makeLenses ''T
+
+-- useless with debrunj indincies ☹
```
Comment (cwgoes) [inline src/Juvix/Core/IR/Subst.hs:33]:
```suggestion
-- useless with de bruijn indices ☹
```

File: src/Juvix/Core/IR/Subst.hs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+-- - take an environment of some kind
+--   + This env should have meta data at some point to better make
+--     decisions. Or better yet tag it to the structure itself in some
+--     extension.
+-- - Make this run on some type of size and usage measurement.
+--   + The current metric just checks if the usage is one. However
+--     this is bad, as we could move the evaluation in the middle of
+--     some computation that degrades performance.
+
+inlineP :: Usage.T -> Types.Elim primTy primVal -> Bool
```
Comment (cwgoes) [inline src/Juvix/Core/IR/Subst.hs:99]: Add comment, is this a function used to decide if we will inline something?

File: src/Juvix/Core/IR/Subst.hs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+instance Hashable Key
+
+data T primTy primVal
+  = T
+      { _depth :: Natural,
+        _sub :: Map.T Key (Types.Elim primTy primVal)
+      }
+  deriving (Show)
+
+makeLenses ''T
```
Comment (cwgoes) [inline src/Juvix/Core/IR/Subst.hs:31]: alright... are they really worth it? relatively simple transformations

(I don't mind lenses but we only use them sometimes, it's a bit inconsistent)

File: src/Juvix/Core/IR/Subst.hs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+f _ (Types.PrimTy ty) =
+  Types.PrimTy ty
+f _ (Types.Star uni) =
+  Types.Star uni
+f subst (Types.Pi usage typ body) =
+  Types.Pi usage (f subst typ) (f (over depth succ subst) body)
+f subst (Types.Lam body) =
+  Types.Lam (f (over depth succ subst) body)
+f subst (Types.Let usage bound body)
+  | inlineP usage bound =
```
Comment (cwgoes) [inline src/Juvix/Core/IR/Subst.hs:52]: honestly I find it odd that this function makes inlining decisions, won't that result in a fair bit of duplicate work (checking `inlineP` at least) if we substitute several times over the same term?

**Review by cwgoes — COMMENTED** [review]
Thanks; a few questions.

File: src/Juvix/Core/IR/Subst.hs (hunk 3 above)
Comment (mariari) [reply src/Juvix/Core/IR/Subst.hs:31]: I'm using it for over, which is useful, nicer than using the record updating syntax for a single field.

File: src/Juvix/Core/IR/Subst.hs (hunk 4 above)
Comment (mariari) [reply src/Juvix/Core/IR/Subst.hs:52]: Ideally we'll have annotation extension to this, should it should be better. The GHC inliner does something very similar, as it checks things as it goes.

File: src/Juvix/Core/IR/Subst.hs (hunk 2 above)
Comment (mariari) [reply src/Juvix/Core/IR/Subst.hs:104]: done

**Review by cwgoes — APPROVED** [review]
ACK

**Pull request merged**